service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/SenderLinkHandler.java

```diff
@@ -0,0 +1,190 @@
+    }
+
+    void close()
+    {
+        if (this.senderLink.getLocalState() != EndpointState.CLOSED)
+        {
+            log.debug("Closing {} sender link with link correlation id {}", getLinkInstanceType(), this.linkCorrelationId);
+            this.senderLink.close();
+        }
+    }
+
+    int sendMessageAndGetDeliveryTag(MessageImpl protonMessage)
+    {
+        // Callers of this method are responsible for putting the returned delivery tag into the inProgressMessages map
+        // so that this link can respond to this message being acknowledged appropriately
+
+        //want to avoid negative delivery tags since -1 is the designated failure value
+        if (this.nextTag == Integer.MAX_VALUE || this.nextTag < 0)
+        {
+            this.nextTag = 0;
+        }
+        else
+        {
+            this.nextTag++;
+        }
+
+        byte[] msgData = new byte[1024];
+        int length;
+
+        while (true)
+        {
+            try
+            {
+                length = protonMessage.encode(msgData, 0, msgData.length);
+                break;
+            }
+            catch (BufferOverflowException e)
+            {
+                msgData = new byte[msgData.length * 2];
+            }
+        }
+
+        byte[] deliveryTag = String.valueOf(this.nextTag).getBytes();
+
+        Delivery delivery = this.senderLink.delivery(deliveryTag);
+        try
+        {
+            log.trace("Sending {} bytes over the amqp {} sender link with link correlation id {}", length, getLinkInstanceType(), this.linkCorrelationId);
+            int bytesSent = this.senderLink.send(msgData, 0, length);
+            log.trace("{} bytes sent over the amqp {} sender link with link correlation id {}", bytesSent, getLinkInstanceType(), this.linkCorrelationId);
+
+            if (bytesSent != length)
+            {
+                throw new IOException(String.format("Amqp send operation did not send all of the expected bytes for %s sender link with link correlation id %s, retrying to send the message", getLinkInstanceType(), this.linkCorrelationId));
+            }
+
+            boolean canAdvance = this.senderLink.advance();
+
+            if (!canAdvance)
+            {
+                throw new IOException(String.format("Failed to advance the senderLink after sending a message on %s sender link with link correlation id %s, retrying to send the message", getLinkInstanceType(), this.linkCorrelationId));
+            }
+
+            log.trace("Message was sent over {} sender link with delivery tag {} and hash {}", getLinkInstanceType(), new String(deliveryTag), delivery.hashCode());
+            return Integer.parseInt(new String(deliveryTag));
+        }
+        catch (Exception e)
+        {
+            log.warn("Encountered a problem while sending a message on {} sender link with link correlation id {}", getLinkInstanceType(), this.linkCorrelationId, e);
+            this.senderLink.advance();
+            delivery.free();
```

Comment:
Should we do delivery.free() in a finally block instead?

Reply:
freeing the delivery as the sender prevents the message from being sent, so we only want to do that in the exception case.